Shipping Integration - Sync shipping rates for subdivisions of a country

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #1231.

This PR implements the system for parsing and syncing the suggested shipping rates for states/provinces and postal codes fetched from WC Shipping Zones (provided by #1341) with Google Merchant Center.

Two new adapter classes are created for mapping the store's shipping configuration to Google's `ShippingSettings`:

- **`DBShippingSettingsAdapter`**: Maps the rates obtained from the database to a ShippingSettings object that can be sent to the Google API. It accepts a `db_rates` argument that contains the rates in the same format as set in the `gla_shipping_rates` table.
- **`WCShippingSettingsAdapter`**: Maps the `LocationRate` objects fetched by parsing the WC Shipping Zone (by the `ShippingZone` class). It accepts a `rates_collections` parameter, which is an array of `CountryRatesCollection` containing location rates for each country supported by the store.

We group the shipping rates by their country and applicable shipping area first. An applicable area can be a whole country, a state, or specific postcodes of a country. Then for each group, we separate the rates that have any minimum order amount constraints (specifically free shipping rates with a min order amount threshold).

After the rates are all grouped together, we map them according to their applicable shipping area:

- If the rates only apply to a country, then we create a simple RateGroup with just one flat rate for the whole country.
- If the rates only apply to a state, then we create a table rate group and set the shipping rates for each state by mapping the rate to their associated location ID.
- If the rates apply to postcodes, we again create a table rate group for that country and set the rates for each group of postcodes.

An important note is that Google assumes all postal codes are unique within a country; therefore, the API doesn't accept different shipping rates associated with the same postcode but in different states/provinces. WooCommerce, on the other hand, allows the user to set different postcode values for any country+state combinations so there's a slight difference in how we handle that here. This shouldn't be a problem for us since we only support a handful of countries at the moment and AFAIK all those countries have unique postcodes. However, if we later add support for countries that don't adhere to this rule (like France) we (and Google) might need to change how this mapping works.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Create multiple types of shipping settings, rates, and zones in WooCommerce settings
2. Edit your free listing program or go through the plugin onboarding and select the "Recommended" option for syncing the shipping settings
3. Complete setup and confirm in MC that all your shipping settings are synced (only the supported countries and methods will be synced as mentioned in #1228)
4. Edit your free listing program again and this time choose the "My shipping settings are simple." option
5. Enter shipping rates for the countries and complete setup
6. Check in MC that the shipping rates are now updated to reflect the new ones set in the previous step.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Add - Sync the shipping rates for states/provinces and postal codes to Merchant Center.

Commit:
Create AbstractRateGroupAdapter

Commit:
Create CountryPostcodesRateGroupAdapter

Commit:
Create unit tests for CountryPostcodesRateGroupAdapter

Commit:
Create StatesPostcodesRateGroupAdapter

Commit:
Create unit tests for StatesPostcodesRateGroupAdapter

Commit:
Create StatesRateGroupAdapter

Commit:
Create unit tests for StatesRateGroupAdapter

Commit:
Create PostcodeRange

Commit:
Add unit tests for PostcodeRange

Commit:
Refactor Location to ShippingLocation

Uses the new PostcodeRange class for storing postcodes. Adds the Google ID to each location and new method to get applicable shipping area.

Commit:
Separate ShippingRate objects per shipping class

Instead of returning one shipping rate object and including the rates of each class under the `$shipping_class_rates` property of the `ShippingRate` class, we now return multiple `ShippingRate` objects each with its own set of `applicable_classes`

Commit:
Group shipping rates using the same shipping class

Commit:
Ignore shipping class rates when returning suggestions

Commit:
Create LocationRatesCollection

Commit:
Create CountryRatesCollection

This class stores the collection of location rates belonging to a single country.

Commit:
Add unit tests for CountryRatesCollection

Commit:
Create ServiceRatesCollection

This class holds the collection of location rates belonging to a Merchant Center shipping service.

Commit:
Add unit tests for ServiceRatesCollection

Commit:
Create AbstractShippingSettingsAdapter

Commit:
Create DBShippingSettingsAdapter

This class maps the simple shipping rates obtained from the database (previously stored by the shipping API) to Merchant Center shipping services and settings.

Commit:
Add unit tests for DBShippingSettingsAdapter

Commit:
Create WCShippingSettingsAdapter

This class maps the location rates generated by parsing the WooCommerce store shipping zone configurations to Merchant Center shipping settings and services.

Commit:
Add unit tests for WCShippingSettingsAdapter

Commit:
Sync the shipping rates to MC using the adapter classes

nima-karimi commented:
Still a work-in-progress because I have hit an error when trying to sync the shipping rates that belong to specific postcodes of a state and forwarded the following question to the Google team:

```
I’m hitting the following error from the Google shipping API when I set a [table](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#Table) for a RateGroup containing a ‘locationIdSets’ header for its rows and ‘postalCodeGroupNames` for the columns:

[services[4].rateGroups[0].mainTable] Both dimensions of the table are of type delivery location.

This happens when, for example, we set the following table:

+----------------------------------------+
Location ID \ Postcodes    ||  1001 - 1009
123 (CA)                             ||  20
456 (NV)                            ||  10
+----------------------------------------+

I assume this means that we can’t use locationIds AND postal codes in the same table but there are valid use cases for this, for example, when the user sets a shipping rate for just a few postcodes of a certain state/province.

This constraint probably comes from the assumption that all postcodes are unique within a country and we should simply set them on a country level instead of on a state/province level. However, this is not true for all countries and there’s no such constraint defined in the WooCommerce shipping settings.

The question is: should we just assume that all postcodes are unique within a country and simply set them on a country level? Since we currently only support a select number of countries, this might be feasible but it’s something we might need to change later once we support more countries.
```

I'll update the PR description with more details once the above issue is resolved and this is ready for a review.

Commit:
Remove unused class methods

Commit:
Map locations with both states and postcodes using only postcodes

This way we are assuming that all postcodes are unique values within any of the supported countries. Since Google does not support two-dimensional RateGroup tables with both dimensions of location type, we need to create one-dimensional tables with only the postcodes and their associated rates for a country.

nima-karimi commented:
This PR is now unblocked and ready for review. The Google team responded to my question above and their requirements are as I suspected. The summary is: Postcodes are unique within a country; therefore, a rate group cannot define rates for both states and postcodes in the same table.

Since we only support a handful of countries, the above constraint is fine for now. It might bite us back when we add support for more countries that don't adhere to this rule. That said, there's nothing we can do now since Google API is enforcing this constraint on their side.

To make the mapping code compatible with the above, I've ignored the `state` parameter for a shipping location if the location already has `postcodes` defined. This means that if the user defines shipping rates for "US_CA::90210", we set the rate for "US::90210" by creating a table rate.

Commit:
Change method return comment

Commit:
Merge branch 'feature/change-shipping-rate-api' into feature/sync-shipping-rates-for-subdivisions

# Conflicts:
#	src/API/Google/Settings.php

mikkamp commented on src/Shipping/CountryRatesCollection.php:
```diff
@@ -0,0 +1,108 @@
+			// Group rates by their applicable country and affecting shipping area
+			$service_key = $country . $shipping_area;
+			// If the rate has a min order amount constraint, then it should be under a new service
+			if ( ! is_null( $min_order_amount ) ) {
```
Wouldn't it be safer to just use `if ( $min_order_amount )` or do we expect this value to never be set to 0? Or since it's a float compare for it to be larger than 0.0? In WC it's valid to set the minimum order amount to `0` or `0.0`.

Same comment for the function `has_min_order_amount`.

nima-karimi replied:
My main reason for this change was that comparing floats usually [can't be trusted](https://www.php.net/manual/en/language.types.float.php) as they work differently on different systems. So PHP might think 0.001 is 0. That might still be fine given the context I guess but I think simply checking whether the `min_order_amount` has been set or not would work here because even if we get two rates with min_order_amount of 0, it'd be a result of user misconfiguration and not a mapping error.

mikkamp commented on tests/Unit/Shipping/GoogleAdapter/DBShippingSettingsAdapterTest.php:
```diff
@@ -0,0 +1,182 @@
+
+			if ( 'US' === $service->getDeliveryCountry() ) {
+				$this->assertEquals( 'USD', $service->getRateGroups()[0]->getSingleValue()->getFlatRate()->getCurrency() );
+				$this->assertEquals( 10, $service->getRateGroups()[0]->getSingleValue()->getFlatRate()->getValue() );
```
Since we have different delivery times per country, why aren't we asserting here that those are correct as well?

nima-karimi replied:
Done in https://github.com/woocommerce/google-listings-and-ads/pull/1455/commits/c0d1a954a3d9ddcc8715e32cfd09b30ebbcdf7f0

mikkamp commented on tests/Unit/Shipping/GoogleAdapter/DBShippingSettingsAdapterTest.php:
```diff
@@ -0,0 +1,182 @@
+		$this->assertEmpty(  $settings->getServices() );
+	}
+
+	public function test_sets_free_shipping_threshold_on_free_rates() {
```
Should we also test free shipping without any threshold?

nima-karimi replied:
The mapping logic would be the same as any other rate but for completeness, it might be a good idea. Added in https://github.com/woocommerce/google-listings-and-ads/pull/1455/commits/c67fcd98b9df0b862081568183c73aeed49b6615

mikkamp commented on tests/Unit/Shipping/PostcodeRangeTest.php:
```diff
@@ -0,0 +1,33 @@
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\Shipping
+ */
+class PostcodeRangeTest extends UnitTest {
```
It looks like each postcode is mapped through this range class but we aren't actually testing ranges. Should we be testing single postcode ranges here as well? Or what about a postcode containing a wildcard?

nima-karimi replied:
Hmm. Do you mean we should test the `set_` and `get_` methods to see whether they can accept and return single postcodes and wildcards?

Given that the PostcodeRange class has limited logic and is more used as a value object, we're only testing its string conversion functionalities. The `set_` and `get_` and construct functionalities are also being tested as part of the string casting tests. A postcode containing a wildcard would still be treated as any other string since this class doesn't handle any formating.

mikkamp replied:
The comment was more intended for the `PostcodesRateGroupAdapterTest` where it only tests mapping with a postcode range. Which I suppose is fine since `PostcodeRangeTest`, tests a single range. I was also a bit confused about why it tests a postcode range for `DE` (I was under the impression that subdivisions was only supported for some countries). I guess it's just the API that has this restriction, and the mapping should work regardless.

I was originally under the impression that the class `Google\Service\ShoppingContent\PostalCodeRange` would also handle syntax checking, but it seems it just allows any value for begin and end, and it's only when it's passed to the API that it might complain about invalid syntax. I was looking for the limitations specified [here](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#PostalCodeRange), while both begin and end can be a string it mentions specific formatting.

I mainly say this is missing, because WooCommerce does not prevent you from saving a postcode rate like `1234?`, Internally it will never match that location when calculating shipping. However I tested that with the mapping and it doesn't complain since it accepts all strings, but when we try to sync that to Google it responds with the following error:
```
{
	"code": "gla_setting_sync_error",
	"message": "{\n  \"error\": {\n    \"code\": 400,\n    \"message\": \"[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.\",\n    \"errors\": [\n      {\n        \"message\": \"[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.\",\n        \"domain\": \"global\",\n        \"reason\": \"invalid\"\n      }\n    ]\n  }\n}\n",
	"data": {
		"status": 500,
		"message": "",
		"data": {
			"error": {
				"code": 400,
				"message": "[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.",
				"errors": [
					{
						"message": "[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.",
						"domain": "global",
						"reason": "invalid"
					}
				]
			}
		}
	}
}
```

mikkamp reviewed (commented):
That's a lot of changes, at initial glance it looks like the code is functioning as expected without any errors.

It's not showing up as one of the changes in this PR. But the function `LocationRatesProcessorTest::setUp` should have a void return type to be compatible with the PHP 8 tests.

Maybe not part of this PR, but I noticed that ShippingRateSuggestionsController no longer uses the ShippingZone parameter.

I tested some different scenarios with various zones/rates and they appear to be coming across as expected. For the US I have a mix of both rates for States and for postcode ranges. When I call the funciton `generate_shipping_settings` it creates an instance of `Google\Service\ShoppingContent\Table` with the same postcode ranges:

```
["postalCodeGroupNames"]=>array(1) {
    [0]=>string(30) "US - 99501...99950,1234*,23456"
}

...

["postalCodeRanges"]=>
      array(3) {
        [0]=>
        object(Google\Service\ShoppingContent\PostalCodeRange)#17084 (5) {
          ["postalCodeRangeBegin"]=>string(5) "99501"
          ["postalCodeRangeEnd"]=>string(5) "99950"
        }
        [1]=>
        object(Google\Service\ShoppingContent\PostalCodeRange)#17085 (5) {
          ["postalCodeRangeBegin"]=>string(5) "1234*"
          ["postalCodeRangeEnd"]=>NULL
        }
        [2]=>
        object(Google\Service\ShoppingContent\PostalCodeRange)#17086 (5) {
          ["postalCodeRangeBegin"]=>string(5) "23456"
          ["postalCodeRangeEnd"]=>NULL
        }
      }
```

But I don't see these appear in my Merchant Account:

![image](https://user-images.githubusercontent.com/11388669/166716027-7e8f8d7a-a132-48da-b600-30240f3e268a.png)

Do I need to test postcode ranges separately without rates for zones? Or is there something else I'm missing?

It seems like all those formats should be supported:
![image](https://user-images.githubusercontent.com/11388669/166720311-78b7378e-0b40-46b6-8993-c0b6754081b9.png)

mikkamp commented:
Update: It did sync the postcode as a separate table, I just failed to notice that it was only showing the first 5 services.

![image](https://user-images.githubusercontent.com/11388669/166727230-1249c322-0f45-4d9e-a036-d939a2ac3075.png)

![image](https://user-images.githubusercontent.com/11388669/166727445-ab385611-26fa-4328-b8ed-dbf199095cfe.png)

Commit:
Change test `setUp` method return type

Commit:
Merge branch 'feature/change-shipping-rate-api' into feature/sync-shipping-rates-for-subdivisions

Commit:
Use method to check `has_min_order_amount`

Commit:
Fix method comment

Commit:
Test delivery times

Commit:
Test no threshold free shipping mapping

nima-karimi commented:
Thanks, @mikkamp 🙏 And sorry for the large changeset. I've responded to your comments and made some changes to the code. Ready for another review.

> It's not showing up as one of the changes in this PR. But the function LocationRatesProcessorTest::setUp should have a void return type to be compatible with the PHP 8 tests.

Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1455/commits/a309d6754a5e9ffbf1c015ac7414a2bd00ceb3f5

 > Maybe not part of this PR, but I noticed that ShippingRateSuggestionsController no longer uses the ShippingZone parameter.

This was a change introduced in #1422. I've fixed it in there and merged the commit here.

mikkamp commented:
Thanks for those extra changes, that looks good and it's passing the unit tests now.

I just had one worry left, as mentioned [here](https://github.com/woocommerce/google-listings-and-ads/pull/1455#discussion_r865653189). I just think we shouldn't make an error like that prevent syncing from happening, especially since the sync errors aren't surfaced in the UI, they would have to check the logs to find out what to fix.

Commit:
Change test case country to a supported one

nima-karimi commented:
> The comment was more intended for the PostcodesRateGroupAdapterTest where it only tests mapping with a postcode range. Which I suppose is fine since PostcodeRangeTest, tests a single range.

Yes, since that functionality is not defined in the `PostcodesRateGroupAdapter` class it would make more sense to be tested in the `PostcodeRangeTest`.

> I was also a bit confused about why it tests a postcode range for DE (I was under the impression that subdivisions was only supported for some countries). I guess it's just the API that has this restriction, and the mapping should work regardless.

We don't have any checks in the code since we automatically filter only the supported countries when we build the mapping and the list of suggested shipping countries won't come from external/user source. I changed that country to `AU` for the time being but as you said it wouldn't have any effect on the code since that list is dynamic and can be changed later.

> I was originally under the impression that the class Google\Service\ShoppingContent\PostalCodeRange would also handle syntax checking, but it seems it just allows any value for begin and end, and it's only when it's passed to the API that it might complain about invalid syntax. I was looking for the limitations specified [here](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#PostalCodeRange), while both begin and end can be a string it mentions specific formatting.
> I mainly say this is missing, because WooCommerce does not prevent you from saving a postcode rate like 1234?, Internally it will never match that location when calculating shipping. However I tested that with the mapping and it doesn't complain since it accepts all strings, but when we try to sync that to Google it responds with the following error
 
Yes, we're merely passing the values from WC to MC API and no validation happening here. This is kind of a trade-off, whether we want to add the same rules defined by Google API in our own code or simply let Google handle the validation and relay that back to the user. I decided to go with the API way because it would make our code simpler. Do you think checking them locally would be more beneficial and worth the extra logic? We could argue that it reduces the # of invalid requests but this isn't like product sync and this endpoint is only called once by user action.

mikkamp reviewed (approved):
> Yes, we're merely passing the values from WC to MC API and no validation happening here. This is kind of a trade-off, whether we want to add the same rules defined by Google API in our own code or simply let Google handle the validation and relay that back to the user. I decided to go with the API way because it would make our code simpler. Do you think checking them locally would be more beneficial and worth the extra logic? We could argue that it reduces the # of invalid requests but this isn't like product sync and this endpoint is only called once by user action.

If the UI would return the error I think it would be easier for them to correct and resync, but all they would get to see is "Unable to complete your setup. Please try again later.", for the actual error they would need to check the logs. It was just on my mind that we ignore other rates (for example it uses a formula for the rate), so it would seem logical to ignore the postcode line.

Granted though, this won't happen often, so I'll leave it up to you to decide. We could also move it to a low priority issue, so I'll go ahead and approve the PR since I didn't come across any other blockers.

nima-karimi commented:
Thanks for the thorough review, @mikkamp 🙏 

> If the UI would return the error I think it would be easier for them to correct and resync, but all they would get to see is "Unable to complete your setup. Please try again later.", for the actual error they would need to check the logs. 

This is a valid point and a nice improvement that applies to the rest of the API endpoints as well. It's just, as you mentioned, beyond the scope of this PR. I think we can work on improving the error handling and relaying the API errors in a separate PR.
